### PR TITLE
Target base scale support

### PR DESCRIPTION
**Pull request by alrod**
### Issue describing the changes in this PR

[Target base scale in runtime driven scale](https://github.com/Azure/azure-webjobs-sdk/issues/2909)

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

The change adds target base scale support for function runtime.

### CODE REVIEW

**Comment by fabiocav** [comment]
@alrod can you please rebase this PR? Thanks!

**Comment by alrod** [comment]
> @alrod can you please rebase this PR? Thanks!

rebased

File: src/WebJobs.Script/Scale/FunctionsScaleManager.cs [hunk 1]
```diff
@@ -66,10 +66,18 @@ public virtual async Task<ScaleStatusResult> GetScaleStatusAsync(ScaleStatusCont
...
+                        int vote = 0;
+                        if (scaleStatus.TargetWorkerCount.HasValue)
+                        {
+                            vote = scaleStatus.TargetWorkerCount.Value;
+                        }
+                        else
+                        {
+                            vote = Convert.ToInt32(scaleStatus.Vote);
+                        }
+                        _logger.LogDebug($"Monitor '{monitor.Descriptor.Id}' voted '{vote}'");
```
Comment (liliankasem) [inline src/WebJobs.Script/Scale/FunctionsScaleManager.cs:79]: structured logging

```suggestion
                        _logger.LogDebug("Monitor '{id}' voted '{vote}'", monitor.Descriptor.Id, vote);
```

File: src/WebJobs.Script/WebJobs.Script.csproj [hunk 2]
```diff
@@ -47,7 +47,7 @@
     <PackageReference Include="Microsoft.ApplicationInsights.DependencyCollector" Version="2.20.0" />
     <PackageReference Include="Microsoft.ApplicationInsights.WindowsServer" Version="2.20.0" />
     <PackageReference Include="Microsoft.ApplicationInsights.WindowsServer.TelemetryChannel" Version="2.20.0" />
-    <PackageReference Include="Microsoft.Azure.WebJobs" Version="3.0.34-11957" />
+    <PackageReference Include="Microsoft.Azure.WebJobs" Version="3.0.34-11967" />
```
Comment (liliankasem) [inline src/WebJobs.Script/WebJobs.Script.csproj:50]: nit: extra space
```suggestion
    <PackageReference Include="Microsoft.Azure.WebJobs" Version="3.0.34-11967" />
```

**Review by liliankasem — APPROVED** [review]

File: src/WebJobs.Script/Scale/FunctionsScaleManager.cs [hunk 3]
```diff
@@ -66,10 +66,18 @@ public virtual async Task<ScaleStatusResult> GetScaleStatusAsync(ScaleStatusCont
                             WorkerCount = context.WorkerCount,
                             Metrics = metrics
                         };
-                        var result = monitor.GetScaleStatus(scaleStatusContext);
-
-                        _logger.LogDebug($"Monitor '{monitor.Descriptor.Id}' voted '{result.Vote.ToString()}'");
-                        votes.Add(result.Vote);
+                        var scaleStatus = monitor.GetScaleStatus(scaleStatusContext);
+                        int vote = 0;
```
Comment (jviau) [inline src/WebJobs.Script/Scale/FunctionsScaleManager.cs:70]: Null-coalesce would make this more concise:
```suggestion
                        int vote = scaleStatus.TargetWorkerCount ?? Convert.ToInt32(scaleStatus.Vote);
```

File: src/WebJobs.Script/Scale/FunctionsScaleManager.cs [hunk 4]
```diff
@@ -84,38 +92,36 @@ public virtual async Task<ScaleStatusResult> GetScaleStatusAsync(ScaleStatusCont
...
 
-            return new ScaleStatusResult
-            {
-                Vote = vote
-            };
+            return aggregatedVote;
         }
 
-        internal static ScaleVote GetAggregateScaleVote(List<ScaleVote> votes, ScaleStatusContext context, ILogger logger)
+        internal static int GetAggregateScaleVote(List<int> votes, ScaleStatusContext context, ILogger logger)
```
Comment (jviau) [inline src/WebJobs.Script/Scale/FunctionsScaleManager.cs:100]: I understand this pattern predates your PR, so address this comment at your discretion.

I am not a big fan of functions like this being marked `internal` to unit test them. It makes the class more rigid to refactors later on (as what should be completely opaque implementation details is now being verified by a test) and also allows others to take improper dependencies on this.

Do the unit tests for the public API surface of this class not sufficiently cover this code? If not, can they be adjusted to cover it? If covering this via public API surface is not possible, then maybe this helper is doing more than is necessary?

File: src/WebJobs.Script/Scale/FunctionsScaleManager.cs [hunk 5]
```diff
@@ -84,38 +92,36 @@ public virtual async Task<ScaleStatusResult> GetScaleStatusAsync(ScaleStatusCont
...
-                    vote = ScaleVote.ScaleOut;
+                    vote = votes.Max(); // scale out on max vote
                 }
-                else if (context.WorkerCount > 0 && votes.All(p => p == ScaleVote.ScaleIn))
+                else if (context.WorkerCount > 0 && votes.All(p => p < 0))
                 {
                     // scale in only if all monitors vote scale in
                     logger.LogDebug("Scaling in based on votes");
-                    vote = ScaleVote.ScaleIn;
+                    vote = votes.Where(x => x != 0).Max(); // scale in on max vote
```
Comment (jviau) [inline src/WebJobs.Script/Scale/FunctionsScaleManager.cs:117]: Can the `.Where(x => x != 0)` be removed? The `else if` statement above has a restriction of all votes are less that 0, so no vote here will be 0. Or does `voles.All(p => p < 0)` need to change instead?

File: src/WebJobs.Script/Scale/FunctionsScaleManager.cs [hunk 6]
```diff
@@ -37,11 +37,11 @@ public FunctionsScaleManager(IScaleMonitorManager monitorManager, IScaleMetricsR
         /// </summary>
         /// <param name="context">The context to use for the scale decision.</param>
         /// <returns>The scale vote.</returns>
-        public virtual async Task<ScaleStatusResult> GetScaleStatusAsync(ScaleStatusContext context)
+        public virtual async Task<int> GetScaleStatusAsync(ScaleStatusContext context)
```
Comment (jviau) [inline src/WebJobs.Script/Scale/FunctionsScaleManager.cs:40]: I am more of a fan of the previous explicit result type. The `int` is a bit vague. What does it mean when I return something outside of -1, 0, 1? What unit does this int represent? Maybe just get rid of the wrapper object and use `ScaleVote` enum directly? Unless you do want to support values outside of -1, 0, 1. If that is true, can we think about what the unit represents? Maybe documentation is all that is needed.

Comment (liliankasem) [reply src/WebJobs.Script/Scale/FunctionsScaleManager.cs:40]: I was also curious why we're getting rid of `ScaleStatusResult` for `int`

**Comment by alrod** [comment]
Closing in favor https://github.com/Azure/azure-functions-host/pull/8780

**Pull request closed without merging**